staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"fmt"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+)
+
+// getCRDSchemaForVersion returns the validation schema for given version in given CRD.
+func getCRDSchemaForVersion(crd *apiextensions.CustomResourceDefinition, version string) (*apiextensions.CustomResourceValidation, error) {
+	for _, v := range crd.Spec.Versions {
+		if version != v.Name {
+			continue
+		}
+		if hasPerVersionSchema(crd.Spec.Versions) {
```

Comment:
you can do this check earlier before for loop and return crd.Spec.Validation as early as possible then you will have less nested blocks.

Reply:
agree that unnesting like this would improve readability. up to you if you want to unnest like this:

```go
if !hasPerVersionSchema(crd.Spec.Versions) {
  return crd.Spec.Validation, nil
}
if crd.Spec.Validation != nil {
  return nil, fmt.Errorf("malformed CustomResourceDefinition %s version %s: top-level and per-version schemas must be mutual exclusive", crd.Name, version)
}
for _, v := range crd.Spec.Versions {
  if version == v.Name {
    return v.Schema, nil
  }
}
return nil, fmt.Errorf("version %s not found in CustomResourceDefinition: %v", version, crd.Name)
```